lib/InlineEntryPointsPass.cpp

```diff
@@ -0,0 +1,94 @@
+// Copyright 2018 The Clspv Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+
+#include "llvm/IR/CallingConv.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+
+#include "clspv/Option.h"
+
+using namespace llvm;
+
+namespace {
+class InlineEntryPointsPass : public ModulePass {
+public:
+  static char ID;
+  InlineEntryPointsPass() : ModulePass(ID) {}
+
+  bool runOnModule(Module &M) override;
+
+private:
+  bool InlineFunctions(Module &M);
+};
+} // namespace
+
+namespace clspv {
+ModulePass *createInlineEntryPointsPass() {
+  return new InlineEntryPointsPass();
+}
+} // namespace clspv
+
+char InlineEntryPointsPass::ID = 0;
+static RegisterPass<InlineEntryPointsPass>
+    X("InlineEntryPointsPass",
+      "Exhaustively inline entry points");
+
+bool InlineEntryPointsPass::runOnModule(Module &M) {
+  bool Changed = false;
+  for (bool local_changed = true; local_changed; Changed |= local_changed) {
+    local_changed = InlineFunctions(M);
+  }
+
+  // Clean up dead functions. This done here to avoid ordering requirements on
+  // inlining.
+  std::vector<Function *> to_delete;
+  for (auto &F : M) {
+    if (F.isDeclaration() || F.getCallingConv() == CallingConv::SPIR_KERNEL)
+      continue;
+
+    if (F.user_empty())
```

Comment:
I would have put this in a convergence loop.  We might visit a leaf function before one of its callees.  In which case the leaf sticks around, incorrectly.

Reply:
This example works and is roughly the same:
```
int func_3(local int *in, int n) { return in[n]; }
int func_2(local int *in, int n) { return func_3(in, n); }
int func_1(local int *in, int n) { return func_2(in, n); }
kernel void kernel_1(local int *in, global int *out, int n) {
  out[n] = func_1(in, n);
}
kernel void kernel_2(local int *in, global int *out, int n) {
  out[n] = func_1(in, n);
}

int foo(local int *in, int n) { return func_3(in, n); }
```

I can't explain the difference. For your example LLVM reports that both inner() and outer() have 2 uses before inlining and 1 afterwards. For my example all the uses reported make sense and the cleanup removes all function except the kernels.

Reply:
I can't explain those stray uses.  Weird.